src/Compilers/Core/Portable/SourceGeneration/GeneratorTimerExtensions.cs

```diff
@@ -0,0 +1,68 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Generic;
+using System.Diagnostics.Tracing;
+using System.Text;
+using Roslyn.Utilities;
+using static Microsoft.CodeAnalysis.CodeAnalysisEventSource;
+
+namespace Microsoft.CodeAnalysis
+{
+    internal static class GeneratorTimerExtensions
+    {
+        public static RunTimer CreateGeneratorDriverRunTimer(this CodeAnalysisEventSource eventSource)
+        {
+            var id = Guid.NewGuid().ToString();
+            var enabled = eventSource.IsEnabled(EventLevel.Informational, Keywords.Performance);
+            if (enabled)
+            {
+                eventSource.StartGeneratorDriverRunTime(id);
+            }
+            return new RunTimer(t => eventSource.StopGeneratorDriverRunTime(t.Ticks, id), enabled);
+        }
+
+        public static RunTimer CreateSingleGeneratorRunTimer(this CodeAnalysisEventSource eventSource, ISourceGenerator generator)
+        {
+            var id = Guid.NewGuid().ToString();
+            var type = generator.GetGeneratorType();
+            var enabled = eventSource.IsEnabled(EventLevel.Informational, Keywords.Performance);
+            if (enabled)
+            {
+                eventSource.StartSingleGeneratorRunTime(type.FullName!, type.Assembly.Location, id);
+            }
+            return new RunTimer(t => eventSource.StopSingleGeneratorRunTime(type.FullName!, type.Assembly.Location, t.Ticks, id), enabled);
+        }
+
+        internal readonly struct RunTimer : IDisposable
+        {
+            private readonly SharedStopwatch _timer;
+            private readonly Action<TimeSpan> _callback;
+            private readonly bool _shouldExecuteCallback;
```

Comment:
Rather than having a `bool` indicating whether to execute the callback, consider moving the check to the callback itself or set the callback to `null` or `timeSpan => { }` for the cases where we're not timing.